File: scripts/lib/deposit.ts
// Instruction encoders for the six deposit-lifecycle instructions: deposit, deposit_core,
// approve_deposit, reject_deposit, return_rejected, return_rejected_core. Account order and arg
// shapes are derived mechanically from
// target/idl/bye_machine.json, cross-checked against each instruction's #[derive(Accounts)]
// struct in programs/bye_machine/src/instructions/deposit/.
//
// Pure and synchronous with respect to I/O, same discipline as admin.ts: every state-dependent
// account (an existing pool's id, the depositor's own token/metadata accounts, the member an
// activation displaces) is a required parameter — these encoders derive this program's own PDAs
// but never read on-chain state themselves. `approve_deposit`'s `displaced` is the *result* of
// `tier-resolve.ts`'s `resolveDisplaced`, computed by the caller before calling this encoder.
//
// `deposit` and `return_rejected` both carry the pNFT `authorization_rules_program` /
// `authorization_rules` pair as Anchor `Option<Account>`s: a `None` is encoded by passing this
// program's own id in that slot (`optionalAddress`), never by omitting the account — omitting
// shifts every later account by one.

import type { Address, Instruction } from "@solana/kit";
import {
  Borsh,
  buildAnchorIx,
  derivePda,
  optionalAddress,
  readonly,
  seedFromPubkey,
  signer,
  writable,
  writableSigner,
} from "./anchor.js";
import {
  ASSOCIATED_TOKEN_PROGRAM,
  MPL_CORE_PROGRAM,
  SYSTEM_PROGRAM,
  SYSVAR_INSTRUCTIONS,
  TOKEN_METADATA_PROGRAM,
  TOKEN_PROGRAM,
} from "../constants.js";

// ── Seeds (bye_machine-specific — anchor.ts stays program-agnostic) ─────────────────────────
// Verbatim against programs/bye_machine/src/common/seeds.rs.
const PROTOCOL_CONFIG_SEED = "protocol";
const POOL_SEED = "pool";
const COLLECTION_SEED = "collection";
const POSITION_SEED = "position";
const POSITION_VAULT_SEED = "vault";
const WALLET_STATS_SEED = "wallet";
const TOP_TIER_SEED = "tier";

function seedFromU16Le(n: number): Uint8Array {
  const b = Buffer.alloc(2);
  b.writeUInt16LE(n);
  return new Uint8Array(b);
}

// ── deposit ──────────────────────────────────────────────────────────────────────────────────
/**
 * No args — `deposit`'s entire wire correctness is its 19-account list and order.
 * `depositorToken`/`metadata`/`masterEdition`/`depositorTokenRecord`/`positionTokenRecord` are
 * required params, not derived: they are PDAs (or plain accounts) of the Token Metadata /
 * associated-token programs, not of `PROGRAM_ID`, and this file derives only this program's own
 * PDAs (`pool`, `position`, `position_vault`, `wallet_stats`), matching admin.ts's convention for
 * accounts this program cannot derive itself.
 */
export async function deposit(params: {
  depositor: Address;
  poolId: number;
  /**
   * The collection the presented asset's metadata declares. Required and not defaulted: the
   * program re-derives the `PoolCollection` record from this pool's key and the record's own
   * `collection`, and the asset's metadata must name that same collection — so a caller that
   * guessed wrong fails on chain rather than silently depositing against another record.
   */
  collection: Address;
  nftMint: Address;
  depositorToken: Address;
  metadata: Address;
  masterEdition: Address;
  /**
   * The five pNFT accounts, **required-present on standard 0 and required-absent on standard 1**
   * (design §4.2). Nullable rather than optional parameters: the program resolves the standard
   * from the asset and then requires exactly this shape, so a caller that omitted the field
   * instead of passing `null` would encode a legacy deposit it never chose. A supplied account on
   * a legacy deposit fails on chain with Anchor's `ConstraintRaw` (2003), a missing one on a pNFT
   * deposit with `AccountNotEnoughKeys` (3005).
   */
  depositorTokenRecord: Address | null;
  positionTokenRecord: Address | null;
  sysvarInstructions: Address | null;
  authorizationRulesProgram: Address | null;
  authorizationRules: Address | null;
}): Promise<Instruction> {
  const {
    depositor,
    poolId,
    collection,
    nftMint,
    depositorToken,
    metadata,
    masterEdition,
    depositorTokenRecord,
    positionTokenRecord,
    sysvarInstructions,
    authorizationRulesProgram,
    authorizationRules,
  } = params;

  const { address: pool } = await derivePda([POOL_SEED, seedFromU16Le(poolId)]);
  const { address: poolCollection } = await derivePda([
    COLLECTION_SEED,
    seedFromPubkey(pool),
    seedFromPubkey(collection),
  ]);
  const { address: position } = await derivePda([
    POSITION_SEED,
    seedFromPubkey(pool),
    seedFromPubkey(nftMint),
  ]);
  const { address: positionVault } = await derivePda([POSITION_VAULT_SEED, seedFromPubkey(position)]);
  const { address: walletStats } = await derivePda([
    WALLET_STATS_SEED,
    seedFromPubkey(pool),
    seedFromPubkey(depositor),
  ]);

  return buildAnchorIx({
    name: "deposit",
    accounts: [
      writableSigner(depositor),
      writable(pool),
      readonly(poolCollection),
      writable(position),
      writable(positionVault),
      writable(walletStats),
      readonly(nftMint),
      writable(depositorToken),
      writable(metadata),
      readonly(masterEdition),
      // A `None` optional is passed as the program's own id, and a program address may not be
      // marked writable — so the placeholder takes the readonly role while the present account
      // keeps the writable one `TransferV1` needs. `exit.ts`'s `depositorUsdc` is the same shape.
      depositorTokenRecord === null
        ? readonly(optionalAddress(depositorTokenRecord))
        : writable(depositorTokenRecord),
      positionTokenRecord === null
        ? readonly(optionalAddress(positionTokenRecord))
        : writable(positionTokenRecord),
      readonly(TOKEN_METADATA_PROGRAM),
      readonly(optionalAddress(authorizationRulesProgram)),
      readonly(optionalAddress(authorizationRules)),
      readonly(optionalAddress(sysvarInstructions)),
      readonly(TOKEN_PROGRAM),
      readonly(ASSOCIATED_TOKEN_PROGRAM),
      readonly(SYSTEM_PROGRAM),
    ],
  });
}

// ── deposit_core ─────────────────────────────────────────────────────────────────────────────
/**
 * `deposit`'s MPL-Core twin: no args and **ten** accounts against the principal's nineteen.
 *
 * Twelve Token Metadata slots are gone and three Core ones take their place. None of the twelve
 * has a Core analogue — an `AssetV1` account *is* the instrument, so no mint derives it, no token
 * account holds it, there is no metadata PDA, no master edition, no token record and no auth-rules
 * pair — and `position_vault` names an address that is never allocated, because MPL Core writes
 * the new owner straight into `AssetV1.owner`.
 *
 * `collection` is required and derived from nothing. It is the collection the **asset** declares
 * through `UpdateAuthority::Collection`, which the program re-reads on chain and pins this slot
 * against; a caller who guesses returns `CollectionNotAdmitted` (6100) rather than depositing
 * against another record. `positionVault` is readonly here, not writable: nothing in this
 * program or in MPL Core writes to it, and marking a never-allocated address writable is how a
 * caller ends up paying to create the account the design exists without.
 */
export async function depositCore(params: {
  depositor: Address;
  poolId: number;
  /** The collection the asset's `update_authority` names — not a caller preference. */
  collection: Address;
  /** The `AssetV1` account itself. There is no mint on this family. */
  asset: Address;
}): Promise<Instruction> {
  const { depositor, poolId, collection, asset } = params;

  const { address: pool } = await derivePda([POOL_SEED, seedFromU16Le(poolId)]);
  const { address: poolCollection } = await derivePda([
    COLLECTION_SEED,
    seedFromPubkey(pool),
    seedFromPubkey(collection),
  ]);
  // `position`'s second seed is the asset, where the Token Metadata twin uses the mint — the
  // same slot in the same derivation, because `Position.nft_mint` stores whichever one names the
  // escrowed instrument.
  const { address: position } = await derivePda([
    POSITION_SEED,
    seedFromPubkey(pool),
    seedFromPubkey(asset),
  ]);
  const { address: positionVault } = await derivePda([POSITION_VAULT_SEED, seedFromPubkey(position)]);
  const { address: walletStats } = await derivePda([
    WALLET_STATS_SEED,
    seedFromPubkey(pool),
    seedFromPubkey(depositor),
  ]);

  return buildAnchorIx({
    name: "deposit_core",
    accounts: [
      writableSigner(depositor),
      writable(pool),
      readonly(poolCollection),
      writable(position),
      readonly(positionVault),
      writable(walletStats),
      writable(asset),
      readonly(collection),
      readonly(MPL_CORE_PROGRAM),
      readonly(SYSTEM_PROGRAM),
    ],
  });
}

// ── approve_deposit ──────────────────────────────────────────────────────────────────────────
/**
 * `weightIndex` is `pool.weight_index` at call time — the caller's responsibility to read, same
 * as `initPool`'s `weightIndex` param. `depositor` is `position.depositor`, needed to derive
 * `wallet_stats`'s PDA — likewise caller-supplied, not read here.
 *
 * `lockUntil` is the Season-0 lock (D-091) — `season.tge_at + 14 days` for a Season-0 deposit, `0`
 * otherwise. **Required and non-defaulted, deliberately.** It sits directly after `observedAt` and
 * both are `i64`, so a default here would let every caller that forgot the lock encode a valid,
 * silently unlocked position — and the program-side pin cannot see a client omission. `0` must be
 * written by the caller who means it.
 *
 * `displaced` carries `PromoteOutcome::Swapped`'s evicted member as a `remaining_accounts` entry
 * matched **by key** (`tier::settle_displaced_member` → `.find()` → `AccountNotEnoughKeys` /
 * 3005 if absent). Required and non-defaulted — resolve it with `tier-resolve.ts`'s
 * `resolveDisplaced(rpc, pool, { kind: "activating", value })` before calling this encoder; a
 * defaulted or optional param here would let a client compile while silently failing every
 * activation that displaces a tier member once the tier fills.
 */
export async function approveDeposit(params: {
  operator: Address;
  poolId: number;
  nftMint: Address;
  depositor: Address;
  weightIndex: Address;
  displaced: Address | null;
  value: bigint;
  observedAt: bigint;
  lockUntil: bigint;
}): Promise<Instruction> {
  const { operator, poolId, nftMint, depositor, weightIndex, displaced, value, observedAt, lockUntil } =
    params;

  const { address: protocolConfig } = await derivePda([PROTOCOL_CONFIG_SEED]);
  const { address: pool } = await derivePda([POOL_SEED, seedFromU16Le(poolId)]);
  const { address: position } = await derivePda([
    POSITION_SEED,
    seedFromPubkey(pool),
    seedFromPubkey(nftMint),
  ]);
  const { address: topTier } = await derivePda([TOP_TIER_SEED, seedFromPubkey(pool)]);
  const { address: walletStats } = await derivePda([
    WALLET_STATS_SEED,
    seedFromPubkey(pool),
    seedFromPubkey(depositor),
  ]);

  return buildAnchorIx({
    name: "approve_deposit",
    // Order is (value, observedAt, lockUntil) — the last two are both i64 and transposing them
    // encodes to the same 24 bytes. See approve_deposit.rs's own pin.
    args: new Borsh().u64(value).i64(observedAt).i64(lockUntil).build(),
    accounts: [
      signer(operator),
      readonly(protocolConfig),
      writable(pool),
      writable(position),
      writable(weightIndex),
      writable(topTier),
      writable(walletStats),
    ],
    remaining: displaced === null ? [] : [writable(displaced)],
  });
}

// ── reject_deposit ───────────────────────────────────────────────────────────────────────────
/**
 * No args beyond `reason` — `position` is a PDA of `[POSITION_SEED, pool, nft_mint]`, seeded
 * directly from `pool`/`nftMint` rather than from a `poolId`: the IDL's own seed path
 * (`position.pool`) is the `Position` account's stored field, which the client supplies as the
 * PDA's own seed material, not something derived from a pool-counter argument.
 */
export async function rejectDeposit(params: {
  operator: Address;
  pool: Address;
  nftMint: Address;
  reason: number;
}): Promise<Instruction> {
  const { operator, pool, nftMint, reason } = params;

  const { address: protocolConfig } = await derivePda([PROTOCOL_CONFIG_SEED]);
  const { address: position } = await derivePda([
    POSITION_SEED,
    seedFromPubkey(pool),
    seedFromPubkey(nftMint),
  ]);

  return buildAnchorIx({
    name: "reject_deposit",
    args: new Borsh().u16(reason).build(),
    accounts: [signer(operator), readonly(protocolConfig), writable(position)],
  });
}

// ── return_rejected ──────────────────────────────────────────────────────────────────────────
/**
 * No args — `return_rejected`'s entire wire correctness is its 17-account list and order.
 * Permissionless: `payer` funds the transaction (`writableSigner`), `depositor` is an
 * `UncheckedAccount` pinned on-chain to `position.depositor` (`writable`, not a signer) — the
 * NFT and both rent refunds land there regardless of who submits the transaction. This differs
 * from `deposit`'s signer shape, where the depositor is the sole signer.
 *
 * `depositorToken` is a bare `UncheckedAccount` here (no mint/owner constraint on-chain), unlike
 * `deposit`'s constrained `depositor_token` — a known open custody question (F-2). This encoder
 * only encodes the account; it does not validate it.
 */
export async function returnRejected(params: {
  payer: Address;
  depositor: Address;
  pool: Address;
  nftMint: Address;
  depositorToken: Address;
  metadata: Address;
  masterEdition: Address;
  /**
   * The five pNFT accounts, **required-present on standard 0 and required-absent on standard 1**
   * (design §4.2 / flows §1.2). Nullable rather than optional parameters: the exit reads the
   * standard off the position and then requires exactly this shape, so a caller that omitted the
   * field instead of passing `null` would encode a legacy release it never chose. A supplied
   * account on a legacy release fails with Anchor's `ConstraintRaw` (2003), a missing one on a
   * pNFT release with `AccountNotEnoughKeys` (3005).
   */
  positionTokenRecord: Address | null;
  depositorTokenRecord: Address | null;
  sysvarInstructions: Address | null;
  authorizationRulesProgram: Address | null;
  authorizationRules: Address | null;
}): Promise<Instruction> {
  const {
    payer,
    depositor,
    pool,
    nftMint,
    depositorToken,
    metadata,
    masterEdition,
    positionTokenRecord,
    depositorTokenRecord,
    sysvarInstructions,
    authorizationRulesProgram,
    authorizationRules,
  } = params;

  const { address: position } = await derivePda([
    POSITION_SEED,
    seedFromPubkey(pool),
    seedFromPubkey(nftMint),
  ]);
  const { address: positionVault } = await derivePda([POSITION_VAULT_SEED, seedFromPubkey(position)]);

  return buildAnchorIx({
    name: "return_rejected",
    accounts: [
      writableSigner(payer),
      writable(depositor),
      writable(position),
      writable(positionVault),
      readonly(nftMint),
      writable(depositorToken),
      writable(metadata),
      readonly(masterEdition),
      // A `None` optional is passed as the program's own id, and a program address may not be
      // marked writable — so the placeholder takes the readonly role while a present account keeps
      // the writable one `TransferV1` needs.
      positionTokenRecord === null ? readonly(optionalAddress(positionTokenRecord)) : writable(positionTokenRecord),
      depositorTokenRecord === null ? readonly(optionalAddress(depositorTokenRecord)) : writable(depositorTokenRecord),
      readonly(TOKEN_METADATA_PROGRAM),
      readonly(optionalAddress(authorizationRulesProgram)),
      readonly(optionalAddress(authorizationRules)),
      readonly(optionalAddress(sysvarInstructions)),
      readonly(TOKEN_PROGRAM),
      readonly(ASSOCIATED_TOKEN_PROGRAM),
      readonly(SYSTEM_PROGRAM),
    ],
  });
}

// ── return_rejected_core ─────────────────────────────────────────────────────────────────────
/**
 * `return_rejected`'s MPL-Core twin: no args and **eight** accounts against the principal's
 * seventeen. Permissionless on exactly the principal's terms — `payer` funds the transaction and
 * `depositor` is an `UncheckedAccount` pinned on-chain to `position.depositor`, so both the card
 * and the position's rent land there whoever submits.
 *
 * Nine Token Metadata slots are gone and one Core slot takes their place, the same trade
 * `depositCore` makes in the other direction: an `AssetV1` account *is* the instrument, so there
 * is no mint, no destination token account, no metadata PDA, no master edition, no token-record
 * pair and no auth-rules pair — and `collection` appears because MPL Core needs the collection
 * account present to run its plugins over the transfer.
 *
 * `positionVault` is readonly, not writable: the escrow identity is never allocated as an
 * account — its only on-chain expression is `AssetV1.owner` — so nothing writes to it, and
 * marking a never-allocated address writable is how a caller ends up paying to create the
 * account the design exists without. `asset` is writable because `TransferV1` rewrites that
 * owner field.
 */
export async function returnRejectedCore(params: {
  payer: Address;
  depositor: Address;
  pool: Address;
  /** The `AssetV1` account itself. There is no mint on this family. */
  asset: Address;
  /** The collection the asset's `update_authority` names — not a caller preference. */
  collection: Address;
}): Promise<Instruction> {
  const { payer, depositor, pool, asset, collection } = params;

  const { address: position } = await derivePda([
    POSITION_SEED,
    seedFromPubkey(pool),
    seedFromPubkey(asset),
  ]);
  const { address: positionVault } = await derivePda([POSITION_VAULT_SEED, seedFromPubkey(position)]);

  return buildAnchorIx({
    name: "return_rejected_core",
    accounts: [
      writableSigner(payer),
      writable(depositor),
      writable(position),
      readonly(positionVault),
      writable(asset),
      readonly(collection),
      readonly(MPL_CORE_PROGRAM),
      readonly(SYSTEM_PROGRAM),
    ],
  });
}

File: programs/bye_machine/src/instructions/deposit/deposit.rs
use anchor_lang::error::ErrorCode;
use anchor_lang::prelude::*;
use anchor_lang::solana_program::program_error::ProgramError;
use anchor_lang::solana_program::sysvar::instructions::ID as SYSVAR_INSTRUCTIONS_ID;
use anchor_spl::associated_token::AssociatedToken;
use anchor_spl::token::{Mint, Token, TokenAccount};
use mpl_token_metadata::ID as TOKEN_METADATA_ID;

use crate::common::errors::ByeMachineError;
use crate::common::escrow::{transfer_pnft, transfer_spl, validate_admission};
use crate::common::events::DepositPending;
use crate::common::seeds::{POOL_SEED, POSITION_SEED, POSITION_VAULT_SEED, WALLET_STATS_SEED};
use crate::common::standards::{STANDARD_LEGACY, STANDARD_PNFT};
use crate::state::{Pool, PoolCollection, Position, PositionState, WalletStats};

pub fn handler(ctx: Context<Deposit>) -> Result<()> {
    require!(
        !ctx.accounts.pool.deposits_paused,
        ByeMachineError::DepositsPaused
    );

    let nft_mint = ctx.accounts.nft_mint.key();

    // The gate resolves the standard from the asset itself rather than assuming one.
    let standard = validate_admission(
        &ctx.accounts.nft_mint.to_account_info(),
        &ctx.accounts.metadata,
        &ctx.accounts.master_edition,
        &ctx.accounts.pool.key(),
        &ctx.accounts.pool_collection.key(),
        &ctx.accounts.pool_collection,
    )?;
    // The five pNFT accounts are **required-present** on standard 0 and **required-absent** on
    // standard 1 — the second half is the one a compile-time pin cannot see, and it is not
    // decorative: a legacy mint deposited with a token record supplied would otherwise leave the
    // account carried, unread, in a transaction whose shape says pNFT.
    //
    // Handler-side rather than `#[account(constraint = ..)]` because the condition is the standard
    // `validate_admission` has only just resolved, and account validation ran before it. Anchor's
    // own codes are reused: `AccountNotEnoughKeys` (3005) for an omission, `ConstraintRaw` (2003)
    // for a supplied account that must be absent.
    match standard {
        STANDARD_PNFT => {
            let depositor_token_record = ctx
                .accounts
                .depositor_token_record
                .as_ref()
                .ok_or(ErrorCode::AccountNotEnoughKeys)?;
            let position_token_record = ctx
                .accounts
                .position_token_record
                .as_ref()
                .ok_or(ErrorCode::AccountNotEnoughKeys)?;
            let sysvar_instructions = ctx
                .accounts
                .sysvar_instructions
                .as_ref()
                .ok_or(ErrorCode::AccountNotEnoughKeys)?;
            transfer_pnft(
                &ctx.accounts.depositor_token.to_account_info(),
                &ctx.accounts.depositor.to_account_info(),
                &ctx.accounts.position_vault.to_account_info(),
                &ctx.accounts.position.to_account_info(),
                &ctx.accounts.nft_mint.to_account_info(),
                &ctx.accounts.metadata,
                &ctx.accounts.master_edition,
                depositor_token_record,
                position_token_record,
                &ctx.accounts.depositor.to_account_info(),
                &ctx.accounts.depositor.to_account_info(),
                &ctx.accounts.system_program.to_account_info(),
                sysvar_instructions,
                &ctx.accounts.token_program.to_account_info(),
                &ctx.accounts.associated_token_program.to_account_info(),
                &ctx.accounts.token_metadata_program,
                ctx.accounts.authorization_rules_program.as_deref(),
                ctx.accounts.authorization_rules.as_deref(),
                &[],
            )?;
        }
        STANDARD_LEGACY => {
            require!(
                ctx.accounts.depositor_token_record.is_none(),
                ErrorCode::ConstraintRaw
            );
            require!(
                ctx.accounts.position_token_record.is_none(),
                ErrorCode::ConstraintRaw
            );
            require!(
                ctx.accounts.sysvar_instructions.is_none(),
                ErrorCode::ConstraintRaw
            );
            require!(
                ctx.accounts.authorization_rules_program.is_none(),
                ErrorCode::ConstraintRaw
            );
            require!(
                ctx.accounts.authorization_rules.is_none(),
                ErrorCode::ConstraintRaw
            );
            transfer_spl(
                &ctx.accounts.depositor_token.to_account_info(),
                &ctx.accounts.position_vault.to_account_info(),
                &ctx.accounts.depositor.to_account_info(),
                &ctx.accounts.token_program.to_account_info(),
                &[],
            )?;
        }
        // `standard` is a `u8`, so the arm is required by the match rather than chosen — the
        // closed set lives in `resolve_standard`, which is where a value outside {0, 1} is
        // actually refused. Unreachable today; it **rejects** rather than falling through,
        // because the alternative to a transfer here is a position opened over an escrow that
        // never happened. `standard == 2` reaches `deposit_core`, never this instruction.
        _ => return Err(ByeMachineError::WrongStandardForInstruction.into()),
    }

    let position_id = ctx.accounts.pool.position_counter;
    ctx.accounts.pool.position_counter = position_id
        .checked_add(1)
        .ok_or(ProgramError::ArithmeticOverflow)?;

    let pool_key = ctx.accounts.pool.key();
    let depositor = ctx.accounts.depositor.key();
    let position_key = ctx.accounts.position.key();
    let bump = ctx.bumps.position;
    let vault_bump = ctx.bumps.position_vault;

    let position = &mut ctx.accounts.position;
    position.pool = pool_key;
    position.depositor = depositor;
    position.nft_mint = nft_mint;
    position.position_id = position_id;
    position.state = PositionState::Pending;
    position.recorded_value = 0;
    position.value_observed_at = 0;
    position.deposit_value = 0;
    position.slot_index = 0;
    position.activated_at = 0;
    position.equal_checkpoint = 0;
    position.accrued = 0;
    position.in_tier = false;
    position.tier_checkpoint = 0;
    position.lock_until = 0;
    position.reject_reason = 0;
    position.standard = standard;
    position.bump = bump;
    position.vault_bump = vault_bump;

    let wallet_stats = &mut ctx.accounts.wallet_stats;
    wallet_stats.pool = pool_key;
    wallet_stats.owner = depositor;
    wallet_stats.bump = ctx.bumps.wallet_stats;

    emit!(DepositPending {
        pool: pool_key,
        slot: Clock::get()?.slot,
        position: position_key,
        depositor,
        nft_mint,
    });

    Ok(())
}

#[derive(Accounts)]
pub struct Deposit<'info> {
    #[account(mut)]
    pub depositor: Signer<'info>,

    #[account(
        mut,
        seeds = [POOL_SEED, pool.pool_id.to_le_bytes().as_ref()],
        bump = pool.bump
    )]
    pub pool: Box<Account<'info, Pool>>,

    /// The admission record for the collection the presented asset declares. Its address is
    /// re-derived in the handler from this pool's key and the collection the *asset's* metadata
    /// names, so neither another pool's record for this collection nor this pool's record for
    /// another collection is accepted.
    pub pool_collection: Box<Account<'info, PoolCollection>>,

    #[account(
        init,
        payer = depositor,
        space = 8 + Position::INIT_SPACE,
        seeds = [POSITION_SEED, pool.key().as_ref(), nft_mint.key().as_ref()],
        bump
    )]
    pub position: Box<Account<'info, Position>>,

    #[account(
        init,
        payer = depositor,
        seeds = [POSITION_VAULT_SEED, position.key().as_ref()],
        bump,
        token::mint = nft_mint,
        token::authority = position
    )]
    pub position_vault: Box<Account<'info, TokenAccount>>,

    #[account(
        init_if_needed,
        payer = depositor,
        space = 8 + WalletStats::INIT_SPACE,
        seeds = [WALLET_STATS_SEED, pool.key().as_ref(), depositor.key().as_ref()],
        bump
    )]
    pub wallet_stats: Box<Account<'info, WalletStats>>,

    pub nft_mint: Box<Account<'info, Mint>>,

    #[account(
        mut,
        constraint = depositor_token.mint == nft_mint.key() @ ByeMachineError::StandardNotAdmitted,
        constraint = depositor_token.owner == depositor.key() @ ByeMachineError::NotPositionOwner
    )]
    pub depositor_token: Box<Account<'info, TokenAccount>>,

    /// CHECK: owner, PDA derivation and contents are asserted by `validate_admission`.
    #[account(mut)]
    pub metadata: UncheckedAccount<'info>,

    /// CHECK: owner, PDA derivation and contents are asserted by `validate_admission`.
    pub master_edition: UncheckedAccount<'info>,

    /// CHECK: derivation is asserted by Token Metadata's own `TransferV1` processor. Required
    /// present on standard 0 and absent on standard 1, which is a condition no account
    /// constraint can express — the standard is not resolved until the handler reads the
    /// metadata.
    #[account(mut)]
    pub depositor_token_record: Option<UncheckedAccount<'info>>,

    /// CHECK: derivation is asserted by Token Metadata's own `TransferV1` processor. Optional on
    /// the same terms as `depositor_token_record`.
    #[account(mut)]
    pub position_token_record: Option<UncheckedAccount<'info>>,

    /// CHECK: pinned to the Token Metadata program this pool's pNFTs are governed by.
    #[account(address = TOKEN_METADATA_ID)]
    pub token_metadata_program: UncheckedAccount<'info>,

    /// CHECK: optional, forwarded to `TransferV1`, which pins it to Token Auth Rules.
    pub authorization_rules_program: Option<UncheckedAccount<'info>>,

    /// CHECK: optional, forwarded to `TransferV1`, which asserts it matches the mint's rule set.
    pub authorization_rules: Option<UncheckedAccount<'info>>,

    /// CHECK: pinned to the instructions sysvar `TransferV1` reads for its CPI guard. Optional on
    /// the same terms as the two token records — SPL `Transfer` has no CPI guard to read it.
    #[account(address = SYSVAR_INSTRUCTIONS_ID)]
    pub sysvar_instructions: Option<UncheckedAccount<'info>>,

    pub token_program: Program<'info, Token>,
    pub associated_token_program: Program<'info, AssociatedToken>,
    pub system_program: Program<'info, System>,
}
